interface/main/main_screen.php

```diff
@@ -90,87 +89,151 @@ function generate_html_end()
 $errormsg = '';
 
 ///////////////////////////////////////////////////////////////////////
-// Begin code to support U2F.
+// Begin code to support U2F and APP Based TOTP logic.
 ///////////////////////////////////////////////////////////////////////
+function generate_auth_form($requests, $errortype, $errormsg) {
+
+    generate_html_start(xl('U2F Key Verification'));
+    echo "<p>\n";
+    echo xlt('Insert your key into a USB port and click the Authenticate button below.');
+    echo " " . xlt('Then press the flashing button on your key within 1 minute.') . "</p>\n";
+    echo "<table><tr><td>\n";
+    echo "<input type='button' value='" . xlt('Authenticate UTF') . "' onclick='doAuth()' />\n";
+    echo "<input type='hidden' name='form_requests' value='" . attr($requests) . "' />\n";
+    echo "</td></tr></table>\n";
+    if ($errormsg && $errortype == "UTF") {
+        echo "<p style='color:red;font-weight:bold'>" . text($errormsg) . "</p>\n";
+    }
+
+    echo '<br /><h3>'.xl('or use').'</h3><br />';
+    echo '<h2>'.xl('App Based TOTP Verification').'</h2>';
+    echo "<p>\n";
+    echo xlt('Enter the code from your authentication application on your device.');
+    echo "<table><tr>";
+    echo "<td><input type='text' name='totp' /></td><td>";
+    echo "<button type='submit'>".xla('Authenticate TOTP')."</button>\n";
+    echo "<input type='hidden' name='form_response' value='true' />\n";
+    echo "</td></tr></table>\n";
+    if ($errormsg && $errortype == "TOTP") {
+        echo "<span style='color:red;font-weight:bold'>" . text($errormsg) . "</span>\n";
+    }
+}
+
 
 $regs = array();          // for mapping device handles to their names
 $registrations = array(); // the array of stored registration objects
 $res1 = sqlStatement(
-    "SELECT a.name, a.var1 FROM login_mfa_registrations AS a " .
-    "WHERE a.user_id = ? AND a.method = 'U2F' ORDER BY a.name",
+    "SELECT a.name, a.method, a.var1 FROM login_mfa_registrations AS a " .
+    "WHERE a.user_id = ? ORDER BY a.name",
```

Comment:
There might later be methods other than U2F or TOTP, so I suggest requiring one of those in this query.